crates/consensus/beacon/src/engine/mod.rs

```diff
@@ -1407,10 +1473,24 @@ where
                     }
                 }
                 Poll::Pending => {
-                    if engine_messages_pending {
-                        // both the sync and the engine message receiver are pending
-                        return Poll::Pending
+                    // no more sync events to process
+                    sync_pending = true;
+                }
+            }
+
+            // check prune events if pipeline is idle and both engine and sync events are pending
+            if this.sync.is_pipeline_idle() && engine_messages_pending & sync_pending {
```

Comment:
why can we only do pruning if the others are pending?

Reply:
the logic is to prioritize processing engine & sync messages by processing them fully first, only then do pruning

Reply:
hmm, I think we can improve this condition in such way:
```suggestion
            if this.sync.is_pipeline_idle() && (this.is_prune_active() || engine_messages_pending & sync_pending) {
```

So, check pruning if:
1. Pipeline is idle
2. Either pruning is running and we need to prioritize checking its events OR no engine and sync messages are pending and we may start pruning

Reply:
right, yep that works